sklearn/tree/tree.py

```diff
@@ -510,6 +515,111 @@ def decision_path(self, X, check_input=True):
+        r_node = prior * self.tree_.impurity
+
+        # find parent node ids and leaves
+        child_l = self.tree_.children_left
+        child_r = self.tree_.children_right
+        parents = np.zeros(shape=n_nodes, dtype=np.intp)
+
+        # Uses uint8 to avoid casting when passing to generate_pruned_tree
+        leaves_in_subtree = np.zeros(shape=n_nodes, dtype=np.uint8)
+
+        stack = [(0, -1)]
+        while len(stack) > 0:
+            node_id, parent = stack.pop()
+            parents[node_id] = parent
+            if child_l[node_id] == child_r[node_id]:
+                leaves_in_subtree[node_id] = 1
+            else:
+                stack.append((child_l[node_id], node_id))
+                stack.append((child_r[node_id], node_id))
+
+        # computes number of leaves in all branches and the overall impurity of
+        # the branch. The overall impurity is the sum of r_node in its leaves.
+        n_leaves = np.zeros(shape=n_nodes, dtype=np.int32)
+        leaf_idicies, = np.where(leaves_in_subtree)
+        r_branch = np.zeros(shape=n_nodes, dtype=DTYPE)
+        r_branch[leaf_idicies] = r_node[leaf_idicies]
+
+        # bubble up values to ancestor nodes
+        for idx in leaf_idicies:
+            cur_R = r_node[idx]
+            while idx != 0:
+                par_idx = parents[idx]
+                r_branch[par_idx] += cur_R
+                n_leaves[par_idx] += 1
+                idx = par_idx
+
+        # non-leaves that can be pruned
+        inner_nodes = np.logical_not(leaves_in_subtree)
+        in_subtree = np.ones(shape=n_nodes, dtype=np.bool)
+
+        cur_alpha = 0
+        while cur_alpha < self.alpha:
+            # If root node is a leaf
+            if not inner_nodes[0]:
+                break
+
+            # computes alpha for subtrees
+            g_node = (r_node - r_branch) / (n_leaves - 1)
+            g_node[~inner_nodes] = np.inf
+
+            # smallest value is the weakest link which is pruned
+            cur_idx = np.argmin(g_node)
+            cur_alpha = g_node[cur_idx]
+
+            # descendants of branch are not in subtree
+            stack = [cur_idx]
+            while len(stack) > 0:
+                idx = stack.pop()
+                inner_nodes[idx] = False
+                leaves_in_subtree[idx] = 0
+                in_subtree[idx] = False
+                n_left, n_right = child_l[idx], child_r[idx]
+                if n_left != n_right:
+                    stack.append(n_left)
+                    stack.append(n_right)
+            leaves_in_subtree[cur_idx] = 1
+
+            # updates number of leaves
+            cur_leaves, n_leaves[cur_idx] = n_leaves[cur_idx], 0
+
+            # computes the increase in r_branch to bubble up
+            R_diff = r_node[cur_idx] - r_branch[cur_idx]
+            r_branch[cur_idx] = r_node[cur_idx]
+
+            # bubble up values to ancestors
+            cur_idx = parents[cur_idx]
+            while cur_idx != _tree.TREE_LEAF:
```

Comment:
It's a bit weird to bubble up to a leaf.

Whatever you're comparing to here should explicitely be the same value as what you used when populating `parents`:

`stack = [(0, -1)]`

I would simply use `while cur_idx != -1:`